API: 'D' and offset.Day always represents calendar day

mroeschke opened the pull request:
- [x] closes #22864
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Commit:
Adjust timeseries.rst

Commit:
Replace CalendarDay with Day; remove Day Tick

Commit:
Revert whatsnew entry

Commit:
Undo some docstrings

Commit:
Add a nanos property to allow Day to interact with timedeltas during depreciation

Commit:
Fix some date_range tests

Commit:
Fix some timezone tests

Commit:
Revert "Revert whatsnew entry"

This reverts commit 16de58de1fa114f1d868b7c6bf88daa228f9584a.

Commit:
Adjust language in whatsnew

Commit:
fix some initial timedelta_tests

Commit:
Remove CalendarDay from some tests

Commit:
fix some offset tests

pep8speaks commented:
Hello @mroeschke! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/indexes/datetimes.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/core/indexes/datetimes.py) !

 - There are no PEP8 issues in the file [`pandas/core/indexes/interval.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/core/indexes/interval.py) !

 - There are no PEP8 issues in the file [`pandas/core/indexes/period.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/core/indexes/period.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/datetimes/test_date_range.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/tests/indexes/datetimes/test_date_range.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/datetimes/test_timezones.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/tests/indexes/datetimes/test_timezones.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/timedeltas/test_timedelta_range.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/tests/indexes/timedeltas/test_timedelta_range.py) !

 - There are no PEP8 issues in the file [`pandas/tests/series/test_timezones.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/tests/series/test_timezones.py) !

 - There are no PEP8 issues in the file [`pandas/tests/test_resample.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/tests/test_resample.py) !

 - There are no PEP8 issues in the file [`pandas/tests/tseries/offsets/test_offsets.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/tests/tseries/offsets/test_offsets.py) !

 - There are no PEP8 issues in the file [`pandas/tests/tseries/offsets/test_ticks.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/tests/tseries/offsets/test_ticks.py) !

 - There are no PEP8 issues in the file [`pandas/tseries/offsets.py`](https://github.com/mroeschke/pandas/blob/6fecd5788c720a32d48d4fcffdd2824fafc144f8/pandas/tseries/offsets.py) !

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Add back Day and make CalendarDay _Day

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Undo resample semantics

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Fix additional tests

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -323,18 +323,16 @@ respected absolute time instead (i.e. ``Day(n)`` and acted identically to ``Time
-See the :ref:`documentation here <timeseries.dayvscalendarday>` for more information.
+:class:`Day` and associated frequency alias ``'D'`` will always respect
+calendar day arithmetic
+(:issue:`22864`, :issue:`20596`, :issue:`16980`, :issue:`8774`)
```
u can add this PR number as well

Commit:
Supress warnings for now

Commit:
Comment out warnings and try another patch

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
fix docstring

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Missing import

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
paramerize test

Commit:
Patch up dst edge case

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Add PR number

mroeschke commented:
One test is failing because of https://github.com/pandas-dev/pandas/issues/23270

https://github.com/pandas-dev/pandas/pull/23278 should fix that test.

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Lint

Commit:
uncomment warnings

Commit:
Dont duplicate code, add statement to check for warning

Commit:
Fix whatsnew

mroeschke commented:
Update: 

In general, I have removed the usage of `CalendarDay` and `'CD'` and replaced `CalendarDay` with `_Day` (to ultimately turn into the new `Day`) and the core tests are passing.

I am slowly introducing `FutureWarnings`, but I anticipating this taking a while. I am chugging through Timedelta warnings currently, but I will also need to introduce `FutureWarnings` to `Day` arithmetic, `.shift`, etc. The ending diff will probably be really large.

mroeschke commented:
@jreback What's our policy about catching warnings in the tests? There are already ~1000 `FutureWarning`s for Timedelta usage with 'D', and I will need to add more `FutureWarnings` for Timestamp arithmetic, shift, etc. Do I need to catch/test them all?

jreback commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -394,6 +395,11 @@ def f(x):
 
 def _generate_regular_range(start, end, periods, offset):
+    # TODO: Remove when _Day replaces Day
+    if hasattr(offset, '_prefix') and offset._prefix == 'D':
```
i am not sure why you think warning on all creations like this is a good idea at all and it’s not what we discussed

this would be a nightmare for us and users

IIRC we need a very selective deprecation

mroeschke replied:
@jorisvandenbossche mentioned https://github.com/pandas-dev/pandas/issues/22864#issuecomment-425366667 that we need deprecation for these 4 points I mentioned https://github.com/pandas-dev/pandas/issues/22864#issue-364655882.

Essentially anywhere that 'D' is used as absolute day needs deprecating which hits a lot of tests.

jreback replied:
those are for arithmetic
you are deprecating creation which seems crazy

mroeschke replied:
We also have to depreciate `timedelta_range(..., freq='D')` since `'D'` will be relative and not absolute in favor of `'24H'`

mroeschke replied:
(This was a cumbersome point for me to tackle when I tried this the first time; that's why I opted for `'CD'` originally)

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
remove all warnings so far:

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Move freq replacement to when needed

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Undo bad merge

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
fix test

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
isort and revert fix

Commit:
xfail resample nonexistent test

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
xfail resample test

Commit:
remove 'CD' inplace of 'D'

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Remove CD in test

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
isort

jorisvandenbossche commented:
@mroeschke I will try to give this (finally, sorry about that) a look. Can you give a short status update? Are there still problems you are encountering, .. ?

jorisvandenbossche commented:
One of the example I used in the discussion is now errorring:

```
In [1]: idx = pd.date_range("2016-10-30", freq='H', periods=4*24, tz='Europe/Helsinki')                                                                                         

In [2]: pd.Series(index=idx).resample('D').count()                                                                                                                              
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-aaf4c03a8f51> in <module>
----> 1 pd.Series(index=idx).resample('D').count()

~/scipy/pandas/pandas/core/resample.py in f(self, _method)
    864 for method in ['count']:
    865     def f(self, _method=method):
--> 866         return self._downsample(_method)
    867     f.__doc__ = getattr(GroupBy, method).__doc__
    868     setattr(Resampler, method, f)

~/scipy/pandas/pandas/core/resample.py in _downsample(self, how, **kwargs)
    990         **kwargs : kw args passed to how function
    991         """
--> 992         self._set_binner()
    993         how = self._is_cython_func(how) or how
    994         ax = self.ax

~/scipy/pandas/pandas/core/resample.py in _set_binner(self)
    170         """
    171         if self.binner is None:
--> 172             self.binner, self.grouper = self._get_binner()
    173 
    174     def _get_binner(self):

~/scipy/pandas/pandas/core/resample.py in _get_binner(self)
    178         """
    179 
--> 180         binner, bins, binlabels = self._get_binner_for_time()
    181         bin_grouper = BinGrouper(bins, binlabels, indexer=self.groupby.indexer)
    182         return binner, bin_grouper

~/scipy/pandas/pandas/core/resample.py in _get_binner_for_time(self)
    979         if self.kind == 'period':
    980             return self.groupby._get_time_period_bins(self.ax)
--> 981         return self.groupby._get_time_bins(self.ax)
    982 
    983     def _downsample(self, how, **kwargs):

~/scipy/pandas/pandas/core/resample.py in _get_time_bins(self, ax)
   1428         # general version, knowing nothing about relative frequencies
   1429         bins = lib.generate_bins_dt64(
-> 1430             ax_values, bin_edges, self.closed, hasnans=ax.hasnans)
   1431 
   1432         if self.closed == 'right':

~/scipy/pandas/pandas/_libs/lib.pyx in pandas._libs.lib.generate_bins_dt64()

ValueError: Values falls after last bin
```

jorisvandenbossche commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1136,6 +1096,7 @@ Deprecations
 - :func:`pandas.types.is_datetimetz` is deprecated in favor of `pandas.types.is_datetime64tz` (:issue:`23917`)
 - Creating a :class:`TimedeltaIndex` or :class:`DatetimeIndex` by passing range arguments `start`, `end`, and `periods` is deprecated in favor of :func:`timedelta_range` and :func:`date_range` (:issue:`23919`)
 - Passing a string alias like ``'datetime64[ns, UTC]'`` as the `unit` parameter to :class:`DatetimeTZDtype` is deprecated. Use :class:`DatetimeTZDtype.construct_from_string` instead (:issue:`23990`).
+- Operations where the offset alias ``'D'`` or :class:`Day` acts as a fixed, absolute duration of 24 hours are deprecated. This includes :class:`Timestamp` arithmetic and all operations with :class:`Timedelta`. :class:`Day` will always represent a calendar day in a future version (:issue:`22864`)
```
Can you give some examples here? 

For example, with the following I don't see a change or warning:

```
In [11]: pd.Timestamp("2012-10-28", tz='Europe/Brussels') + pd.offsets.Day()                                                                                                    
Out[11]: Timestamp('2012-10-28 23:00:00+0100', tz='Europe/Brussels')
```

jorisvandenbossche reviewed (commented):
How do you see the way forward here with `_Day`. Is the idea that `Day` will be replaced by that, after deprecating the behaviours that are different? 
(but I don't see any deprecations right now?)

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

mroeschke commented:
Sorry, I have been primarily working on reverting things and getting the test to pass again by rolling back `CalendarDay`.

**TODO**
Add all the depreciation warnings (I anticipate there being quite a few of these). This needs to be included in the following:

- `Day` tick arithmetic with other `Tick`s, `Timedeltas`, and `DatetimeTZ`
- `DatetimeIndex.shift` (tz-aware only)

**Migration**
The plan is for `_Day` (formerly `CalendarDay`) to just replace `Day` once the prior `Day` behavior is replaced.

**Concern**
During the last dev chat, there was interest for `'D'` to be compatible freq argument/offset with both `Timedelta` and `Datetime`. I don't see a clear way to make this possible without adding a lot of monkeypatching.

Example: `timedelta_range(..., freq='D')`;  `to_offset('D')` will return `_Day` in the future and this offset will need to increment a `Timedelta`, but `_Day` + `Timedelta` is an invalid operation.

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Deprecate Tick arithmetic with Day

Commit:
Merge remote-tracking branch 'upstream/master' into make_D_calendar_day

Commit:
Add and fix test for tick airthmetic deprecation

Commit:
Deprecate Day arithmetic with Timestamptz

mroeschke commented:
Closing in favor of https://github.com/pandas-dev/pandas/pull/24330 for now (reverts behavior to pre-`CalendarDay`). We can revisit this after the 0.24.0 release.
